drv/sprot-api/src/lib.rs

```diff
@@ -333,232 +422,238 @@ impl From<u8> for MsgType {
-pub fn payload_buf(size: Option<usize>, buffer: &[u8]) -> &[u8] {
-    let start = HEADER_SIZE;
-    let end = match size {
-        Some(size) => HEADER_SIZE + size,
-        None => buffer.len() - CRC_SIZE,
-    };
-    match buffer.get(start..end) {
-        Some(buf) => buf,
-        None => panic!(), // Don't come to me with your miniscule buffers.
+    /// Deserialize a hubpack encoded message, `M`, from the wrapped buffer
+    pub fn deserialize_hubpack_payload<M>(
+        &self,
+        rxmsg: &VerifiedRxMsg,
+    ) -> Result<M, SprotError>
+    where
+        M: for<'de> Deserialize<'de>,
+    {
+        let (msg, _) = hubpack::deserialize::<M>(self.payload(rxmsg))?;
+        Ok(msg)
     }
-}
 
-/// Read/write access to the first portion or all of the transmit buffer.
-pub fn payload_buf_mut(size: Option<usize>, buffer: &mut [u8]) -> &mut [u8] {
-    let start = HEADER_SIZE;
-    let end = match size {
-        Some(size) => HEADER_SIZE + size,
-        None => buffer.len() - CRC_SIZE,
-    };
-    match buffer.get_mut(start..end) {
-        Some(buf) => buf,
-        None => panic!(), // Don't come to me with your miniscule buffers.
+    /// Parse the first byte of the protocol, returning an appropriate error
+    /// if necessary.
+    fn parse_protocol(&self) -> Result<(), SprotError> {
+        match Protocol::from(self.buf[0]) {
+            Protocol::Ignore => Err(SprotError::NoMessage),
+            Protocol::Busy => Err(SprotError::RotBusy),
+            Protocol::V1 => Ok(()),
+            _ => Err(SprotError::UnsupportedProtocol),
+        }
     }
 }
 
-/// Finish the message header, compute message CRC and return the
-/// number of bytes to transmit from the underlying buffer.
-/// The payload has to have been placed in the bytes designated by payload_range()
-/// and the payload parameter must be the actual bytes that were used for the payload.
-pub fn compose(
-    msgtype: MsgType,
-    payload_size: usize,
-    buffer: &mut [u8],
-) -> Result<usize, MsgError> {
-    let header = MsgHeader {
-        protocol: Protocol::V1,
-        msgtype,
-        len: u16::try_from(payload_size)
-            .map_err(|_| MsgError::BadMessageLength)?,
-    };
-    let _size = hubpack::serialize(&mut buffer[..HEADER_SIZE], &header)
-        .map_err(|_e| MsgError::Serialization)?;
-    if let Some(msg_bytes) = buffer.get(0..HEADER_SIZE + payload_size) {
-        let crc = CRC16.checksum(msg_bytes);
-        let crc_begin = HEADER_SIZE + payload_size;
-        let end = crc_begin + CRC_SIZE;
-        if let Some(crc_buf) = buffer.get_mut(crc_begin..end) {
-            if let Ok(_n) = hubpack::serialize(crc_buf, &crc) {
-                return Ok(end);
-            } else {
-                return Err(MsgError::Serialization);
-            }
+/// A type indicating that the message header has been parsed and the CRC has
+/// been successfully verified
+pub struct VerifiedRxMsg(pub MsgHeader);
```

Comment:
Same question here: should this borrow `&'a RxMsg` to ensure that no one else can modify it?

Reply:
As above, there's something to this, but I'm worried about lifetime issues. I think some restructuring with an `RxMsg` and `TxMsg` that get consumed to create these values would make this clearer, even if references to the buffers weren't held.... We'd make sure that any places where the buffer gets mutated can't be called unless it has a `TxMsg` or `RxMsg` already, which it won't if it was consumed and the code moved on.